src/test/java/RetryExecutorTest.java

```diff
@@ -0,0 +1,58 @@
+import org.junit.jupiter.api.Test;
+
+import java.io.IOException;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class RetryExecutorTest {
+
+    // Test successful execution on the first attempt
+    @Test
+    void testExecute_SuccessOnFirstAttempt() throws InterruptedException {
+        RetryExecutor executor = RetryExecutor.builder()
+                .maxAttempts(3)
+                .exponentialBackoff(100, 1000)
+                .retryOn(IOException.class)
+                .build();
+
+        String result = executor.execute(() -> "Success");
+        assertEquals("Success", result);
+    }
+
+    // Test retry logic for exception
+    @Test
+    void testExecute_RetryOnException() {
+        RetryExecutor executor = RetryExecutor.builder()
+                .maxAttempts(3)
+                .exponentialBackoff(100, 1000)
+                .retryOn(IOException.class)
+                .build();
+
+        // Vi förväntar oss att ett undantag kastas eftersom alla retries misslyckas
+        assertThrows(RetryExecutor.RetryExhaustedException.class, () -> {
+            executor.execute(() -> {
+                throw new IOException("Failed");
+            });
+        });
+    }
+
+    // Test exhaustion handler
+    @Test
+    void testExecute_OnExhaustion() {
+        RetryExecutor executor = RetryExecutor.builder()
+                .maxAttempts(3)
+                .exponentialBackoff(100, 1000)
+                .retryOn(IOException.class)
+                .onExhaustion(ex -> {
+                    System.out.println("Retry exhausted");
+                    return true;
+                })
+                .build();
+
+        assertThrows(RetryExecutor.RetryExhaustedException.class, () -> {
+            executor.execute(() -> {
+                throw new IOException("Failed");
+            });
+        });
+    }
+}
```

Comment:
Are there more test cases we could add? 
What happens if an exception of a type we shouldn't retry on happens? Will the code stop?
What happens if we use 0 or -1 as maxAttempts argument?
Does we have any default values for these settings? Can we create a test that verifies that our code has been called the right number of times? 

Preferable if we don't have to add Mockito or any other mocking framework as a dependency for our project to verify number of calls and instead create our own test double object in test folder that counts number of calls?